yt_dlp/extractor/playsuisse.py

```diff
@@ -257,15 +272,22 @@ def _extract_single(self, media_data):
             self._merge_subtitles(subs, target=subtitles)
 
         return {
-            'id': media_data['id'],
-            'title': media_data.get('name'),
-            'description': media_data.get('description'),
             'thumbnails': thumbnails,
-            'duration': int_or_none(media_data.get('duration')),
             'formats': formats,
             'subtitles': subtitles,
-            'series': media_data.get('seriesName'),
-            'season_number': int_or_none(media_data.get('seasonNumber')),
-            'episode': media_data.get('name') if media_data.get('episodeNumber') else None,
-            'episode_number': int_or_none(media_data.get('episodeNumber')),
+            **traverse_obj(media_data, {
+                'id': ('id', {str}),
+                'title': ('name', {str}),
+                'description': (('descriptionLong', 'description'), {str}, any),
+                'genres': ('contentTypes', ..., {str}, all, {unpack(join_nonempty, delim='; ')}, filter),
+                'creators': ('directors', ..., {str}, all, {unpack(join_nonempty, delim='; ')}, filter),
+                'cast': ('mainCast', ..., {str}, all, {unpack(join_nonempty, delim='; ')}, filter),
```

Comment:
No, these 3 fields need to be a `list`, not `str`. `location` needs to be `str`, which is why I suggested the `join_nonempty` transformation there but not here. Revert and please re-run tests and update them again
```suggestion
                'genres': ('contentTypes', ..., {str}),
                'creators': ('directors', ..., {str}),
                'cast': ('mainCast', ..., {str}),
```

Reply:
This is what I did initially, but then the tests failed. Here the logs:
```
============================================================================================== test session starts ==============================================================================================
platform linux -- Python 3.12.7, pytest-8.3.5, pluggy-1.5.0 -- /home/user/Downloads/yt-dlp_TEST/yt-dlp/yt-dlp-venv/bin/python3
cachedir: .pytest_cache
rootdir: /home/user/Downloads/yt-dlp_TEST/yt-dlp
configfile: pyproject.toml
plugins: rerunfailures-14.0
collected 6007 items / 6003 deselected / 4 selected                                                                                                                                                             

test/test_download.py::TestDownload::test_PlaySuisse PASSED                                                                                                                                               [ 25%]
test/test_download.py::TestDownload::test_PlaySuisse_1 FAILED                                                                                                                                             [ 50%]
test/test_download.py::TestDownload::test_PlaySuisse_2 FAILED                                                                                                                                             [ 75%]
test/test_download.py::TestDownload::test_PlaySuisse_all PASSED                                                                                                                                           [100%]

=================================================================================================== FAILURES ====================================================================================================
________________________________________________________________________________________ TestDownload.test_PlaySuisse_1 _________________________________________________________________________________________

self = <test.test_download.TestDownload testMethod=test_PlaySuisse_1>

    def test_template(self):
        if self.COMPLETED_TESTS.get(tname):
            return
        self.COMPLETED_TESTS[tname] = True
        ie = yt_dlp.extractor.get_info_extractor(test_case['name'])()
        other_ies = [get_info_extractor(ie_key)() for ie_key in test_case.get('add_ie', [])]
        is_playlist = any(k.startswith('playlist') for k in test_case)
        test_cases = test_case.get(
            'playlist', [] if is_playlist else [test_case])
    
        def print_skipping(reason):
            print('Skipping {}: {}'.format(test_case['name'], reason))
            self.skipTest(reason)
    
        if not ie.working():
            print_skipping('IE marked as not _WORKING')
    
        for tc in test_cases:
            if tc.get('expected_exception'):
                continue
            info_dict = tc.get('info_dict', {})
            params = tc.get('params', {})
            if not info_dict.get('id'):
                raise Exception(f'Test {tname} definition incorrect - "id" key is not present')
            elif not info_dict.get('ext') and info_dict.get('_type', 'video') == 'video':
                if params.get('skip_download') and params.get('ignore_no_formats_error'):
                    continue
                raise Exception(f'Test {tname} definition incorrect - "ext" key must be present to define the output file')
    
        if 'skip' in test_case:
            print_skipping(test_case['skip'])
    
        for other_ie in other_ies:
            if not other_ie.working():
                print_skipping(f'test depends on {other_ie.ie_key()}IE, marked as not WORKING')
    
        params = get_params(test_case.get('params', {}))
        params['outtmpl'] = tname + '_' + params['outtmpl']
        if is_playlist and 'playlist' not in test_case:
            params.setdefault('extract_flat', 'in_playlist')
            params.setdefault('playlistend', test_case.get(
                'playlist_mincount', test_case.get('playlist_count', -2) + 1))
            params.setdefault('skip_download', True)
    
        ydl = YoutubeDL(params, auto_init=False)
        ydl.add_default_info_extractors()
        finished_hook_called = set()
    
        def _hook(status):
            if status['status'] == 'finished':
                finished_hook_called.add(status['filename'])
        ydl.add_progress_hook(_hook)
        expect_warnings(ydl, test_case.get('expected_warnings', []))
    
        def get_tc_filename(tc):
            return ydl.prepare_filename(dict(tc.get('info_dict', {})))
    
        res_dict = None
    
        def match_exception(err):
            expected_exception = test_case.get('expected_exception')
            if not expected_exception:
                return False
            if err.__class__.__name__ == expected_exception:
                return True
            return any(exc.__class__.__name__ == expected_exception for exc in err.exc_info)
    
        def try_rm_tcs_files(tcs=None):
            if tcs is None:
                tcs = test_cases
            for tc in tcs:
                tc_filename = get_tc_filename(tc)
                try_rm(tc_filename)
                try_rm(tc_filename + '.part')
                try_rm(os.path.splitext(tc_filename)[0] + '.info.json')
        try_rm_tcs_files()
        try:
            try_num = 1
            while True:
                try:
                    # We're not using .download here since that is just a shim
                    # for outside error handling, and returns the exit code
                    # instead of the result dict.
                    res_dict = ydl.extract_info(
                        test_case['url'],
                        force_generic_extractor=params.get('force_generic_extractor', False))
                except (DownloadError, ExtractorError) as err:
                    # Check if the exception is not a network related one
                    if not isinstance(err.exc_info[1], (TransportError, UnavailableVideoError)) or (isinstance(err.exc_info[1], HTTPError) and err.exc_info[1].status == 503):
                        if match_exception(err):
                            return
                        err.msg = f'{getattr(err, "msg", err)} ({tname})'
                        raise
    
                    if try_num == RETRIES:
                        raise
    
                    print(f'Retrying: {try_num} failed tries\n\n##########\n\n')
    
                    try_num += 1
                except YoutubeDLError as err:
                    if match_exception(err):
                        return
                    raise
                else:
                    break
    
            if is_playlist:
                self.assertTrue(res_dict['_type'] in ['playlist', 'multi_video'])
                self.assertTrue('entries' in res_dict)
                expect_info_dict(self, res_dict, test_case.get('info_dict', {}))
    
            if 'playlist_mincount' in test_case:
                assertGreaterEqual(
                    self,
                    len(res_dict['entries']),
                    test_case['playlist_mincount'],
                    'Expected at least %d in playlist %s, but got only %d' % (
                        test_case['playlist_mincount'], test_case['url'],
                        len(res_dict['entries'])))
            if 'playlist_count' in test_case:
                self.assertEqual(
                    len(res_dict['entries']),
                    test_case['playlist_count'],
                    'Expected %d entries in playlist %s, but got %d.' % (
                        test_case['playlist_count'],
                        test_case['url'],
                        len(res_dict['entries']),
                    ))
            if 'playlist_duration_sum' in test_case:
                got_duration = sum(e['duration'] for e in res_dict['entries'])
                self.assertEqual(
                    test_case['playlist_duration_sum'], got_duration)
    
            # Generalize both playlists and single videos to unified format for
            # simplicity
            if 'entries' not in res_dict:
                res_dict['entries'] = [res_dict]
    
            for tc_num, tc in enumerate(test_cases):
                tc_res_dict = res_dict['entries'][tc_num]
                # First, check test cases' data against extracted data alone
>               expect_info_dict(self, tc_res_dict, tc.get('info_dict', {}))

test/test_download.py:227: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test/helper.py:276: in expect_info_dict
    expect_dict(self, got_dict, expected_dict)
test/helper.py:206: in expect_dict
    self.fail(message)
E   AssertionError: dictionaries differ (genres, creators, cast)
E       genres: expected str, got list
E       creators: expected str, got list
E       cast: expected str, got list
--------------------------------------------------------------------------------------------- Captured stdout call ----------------------------------------------------------------------------------------------
[debug] Loaded 1857 extractors
[PlaySuisse] Requesting session ID
[PlaySuisse] Submitting username
[PlaySuisse] Submitting password
[PlaySuisse] Logging in
[PlaySuisse] Downloading token
[PlaySuisse] Extracting URL: https://www.playsuisse.ch/detail/2573198
[PlaySuisse] 2573198: Downloading JSON metadata
[PlaySuisse] 2573198: Downloading m3u8 information
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec, channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[info] 2573198: Downloading 1 format(s): HLS-2144
[info] Writing video metadata as JSON to: test_PlaySuisse_1_2573198.info.json
[debug] Invoking hlsnative downloader on "https://stipo2-vod-prod.akamaized.net/stipo/2573198_cd6cb3605509d5d179f5c8cc3e419989/hdntl=exp=1746897831~acl=%2fstipo%2f2573198_cd6cb3605509d5d179f5c8cc3e419989%2f*~data=hdntl,allowedCountries%3dswiss~hmac=79ff5b1e061f2682d01cccafd2af357a8f7bdb7ec5e04b87006b50394f2efabf/index-f6-v1.m3u8"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 749
[download] Destination: test_PlaySuisse_1_2573198.mp4
[download] 100% of  120.80KiB in 00:00:00 at 221.97KiB/s                 
________________________________________________________________________________________ TestDownload.test_PlaySuisse_2 _________________________________________________________________________________________

self = <test.test_download.TestDownload testMethod=test_PlaySuisse_2>

    def test_template(self):
        if self.COMPLETED_TESTS.get(tname):
            return
        self.COMPLETED_TESTS[tname] = True
        ie = yt_dlp.extractor.get_info_extractor(test_case['name'])()
        other_ies = [get_info_extractor(ie_key)() for ie_key in test_case.get('add_ie', [])]
        is_playlist = any(k.startswith('playlist') for k in test_case)
        test_cases = test_case.get(
            'playlist', [] if is_playlist else [test_case])
    
        def print_skipping(reason):
            print('Skipping {}: {}'.format(test_case['name'], reason))
            self.skipTest(reason)
    
        if not ie.working():
            print_skipping('IE marked as not _WORKING')
    
        for tc in test_cases:
            if tc.get('expected_exception'):
                continue
            info_dict = tc.get('info_dict', {})
            params = tc.get('params', {})
            if not info_dict.get('id'):
                raise Exception(f'Test {tname} definition incorrect - "id" key is not present')
            elif not info_dict.get('ext') and info_dict.get('_type', 'video') == 'video':
                if params.get('skip_download') and params.get('ignore_no_formats_error'):
                    continue
                raise Exception(f'Test {tname} definition incorrect - "ext" key must be present to define the output file')
    
        if 'skip' in test_case:
            print_skipping(test_case['skip'])
    
        for other_ie in other_ies:
            if not other_ie.working():
                print_skipping(f'test depends on {other_ie.ie_key()}IE, marked as not WORKING')
    
        params = get_params(test_case.get('params', {}))
        params['outtmpl'] = tname + '_' + params['outtmpl']
        if is_playlist and 'playlist' not in test_case:
            params.setdefault('extract_flat', 'in_playlist')
            params.setdefault('playlistend', test_case.get(
                'playlist_mincount', test_case.get('playlist_count', -2) + 1))
            params.setdefault('skip_download', True)
    
        ydl = YoutubeDL(params, auto_init=False)
        ydl.add_default_info_extractors()
        finished_hook_called = set()
    
        def _hook(status):
            if status['status'] == 'finished':
                finished_hook_called.add(status['filename'])
        ydl.add_progress_hook(_hook)
        expect_warnings(ydl, test_case.get('expected_warnings', []))
    
        def get_tc_filename(tc):
            return ydl.prepare_filename(dict(tc.get('info_dict', {})))
    
        res_dict = None
    
        def match_exception(err):
            expected_exception = test_case.get('expected_exception')
            if not expected_exception:
                return False
            if err.__class__.__name__ == expected_exception:
                return True
            return any(exc.__class__.__name__ == expected_exception for exc in err.exc_info)
    
        def try_rm_tcs_files(tcs=None):
            if tcs is None:
                tcs = test_cases
            for tc in tcs:
                tc_filename = get_tc_filename(tc)
                try_rm(tc_filename)
                try_rm(tc_filename + '.part')
                try_rm(os.path.splitext(tc_filename)[0] + '.info.json')
        try_rm_tcs_files()
        try:
            try_num = 1
            while True:
                try:
                    # We're not using .download here since that is just a shim
                    # for outside error handling, and returns the exit code
                    # instead of the result dict.
                    res_dict = ydl.extract_info(
                        test_case['url'],
                        force_generic_extractor=params.get('force_generic_extractor', False))
                except (DownloadError, ExtractorError) as err:
                    # Check if the exception is not a network related one
                    if not isinstance(err.exc_info[1], (TransportError, UnavailableVideoError)) or (isinstance(err.exc_info[1], HTTPError) and err.exc_info[1].status == 503):
                        if match_exception(err):
                            return
                        err.msg = f'{getattr(err, "msg", err)} ({tname})'
                        raise
    
                    if try_num == RETRIES:
                        raise
    
                    print(f'Retrying: {try_num} failed tries\n\n##########\n\n')
    
                    try_num += 1
                except YoutubeDLError as err:
                    if match_exception(err):
                        return
                    raise
                else:
                    break
    
            if is_playlist:
                self.assertTrue(res_dict['_type'] in ['playlist', 'multi_video'])
                self.assertTrue('entries' in res_dict)
>               expect_info_dict(self, res_dict, test_case.get('info_dict', {}))

test/test_download.py:195: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test/helper.py:276: in expect_info_dict
    expect_dict(self, got_dict, expected_dict)
test/helper.py:206: in expect_dict
    self.fail(message)
E   AssertionError: dictionaries differ (genres, creators)
E       genres: expected str, got list
E       creators: expected str, got list
--------------------------------------------------------------------------------------------- Captured stdout call ----------------------------------------------------------------------------------------------
[debug] Loaded 1857 extractors
[PlaySuisse] Requesting session ID
[PlaySuisse] Submitting username
[PlaySuisse] Submitting password
[PlaySuisse] Logging in
[PlaySuisse] Downloading token
[PlaySuisse] Extracting URL: https://www.playsuisse.ch/detail/1115687
[PlaySuisse] 1115687: Downloading JSON metadata
[download] Downloading playlist: They all came out to Montreux
[PlaySuisse] 1112663: Downloading m3u8 information
[PlaySuisse] 1112661: Downloading m3u8 information
[PlaySuisse] 1112664: Downloading m3u8 information
[PlaySuisse] Playlist They all came out to Montreux: Downloading 3 items of 3
[debug] The information of all playlist entries will be held in memory
[download] Downloading item 1 of 3
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec, channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[info] 1112663: Downloading 1 format(s): HLS-3467
[info] Writing video metadata as JSON to: test_PlaySuisse_2_1112663.info.json
[debug] Invoking hlsnative downloader on "https://stipo2-vod-prod.akamaized.net/stipo/1112663_c2caf368b69b671ddb6984f56b119320/hdntl=exp=1746897834~acl=%2fstipo%2f1112663_c2caf368b69b671ddb6984f56b119320%2f*~data=hdntl,allowedCountries%3dswiss~hmac=48ffc1ad88266d2d2f09d3ad3fa128a2c25912793e45b991977c59820d7f6435/index-f6-v1.m3u8"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 399
[download] Destination: test_PlaySuisse_2_1112663.mp4
[download] 100% of  123.74KiB in 00:00:00 at 211.69KiB/s                 
[download] Downloading item 2 of 3
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec, channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[info] 1112661: Downloading 1 format(s): HLS-3367
[info] Writing video metadata as JSON to: test_PlaySuisse_2_1112661.info.json
[debug] Invoking hlsnative downloader on "https://stipo2-vod-prod.akamaized.net/stipo/1112661_f6c5994870307f2236db15aea1ab7581/hdntl=exp=1746897835~acl=%2fstipo%2f1112661_f6c5994870307f2236db15aea1ab7581%2f*~data=hdntl,allowedCountries%3dswiss~hmac=4dc5c2593cc848ed5a450d925b17960729b414551488d562e89983258f02d75f/index-f5-v1.m3u8"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 368
[download] Destination: test_PlaySuisse_2_1112661.mp4
[download] 100% of  123.74KiB in 00:00:00 at 294.01KiB/s                 
[download] Downloading item 3 of 3
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec, channels, acodec, size, br, asr, proto, vext, aext, hasaud, source, id
[info] 1112664: Downloading 1 format(s): HLS-3268
[info] Writing video metadata as JSON to: test_PlaySuisse_2_1112664.info.json
[debug] Invoking hlsnative downloader on "https://stipo2-vod-prod.akamaized.net/stipo/1112664_3a2209c49d6197ef3b51bcde7f07b42e/hdntl=exp=1746897836~acl=%2fstipo%2f1112664_3a2209c49d6197ef3b51bcde7f07b42e%2f*~data=hdntl,allowedCountries%3dswiss~hmac=a0df201d673c896b8710e6b693e278b6e9feabb0ad7e966a451497f4cc5ed354/index-f6-v1.m3u8"
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 376
[download] Destination: test_PlaySuisse_2_1112664.mp4
[download] 100% of  123.74KiB in 00:00:00 at 307.53KiB/s                 
[download] Finished downloading playlist: They all came out to Montreux
============================================================================================ short test summary info ============================================================================================
FAILED test/test_download.py::TestDownload::test_PlaySuisse_1 - AssertionError: dictionaries differ (genres, creators, cast)
        genres: expected str, got list
        creators: expected str, got list
        cast: expected str, got list
FAILED test/test_download.py::TestDownload::test_PlaySuisse_2 - AssertionError: dictionaries differ (genres, creators)
        genres: expected str, got list
        creators: expected str, got list
================================================================================= 2 failed, 2 passed, 6003 deselected in 15.61s =================================================================================
```

Reply:
remove the `genres`, `creators` and `cast` fields from all of the test dicts, run the tests again, and the test script will give you the correct values to add back to them

Reply:
Ah now I get it. That issue was obvious. Sorry for that!

Tests are successful with these changes:
```
platform linux -- Python 3.12.7, pytest-8.3.5, pluggy-1.5.0 -- /home/user/Downloads/yt-dlp_TEST/yt-dlp/yt-dlp-venv/bin/python3
cachedir: .pytest_cache
rootdir: /home/user/Downloads/yt-dlp_TEST/yt-dlp
configfile: pyproject.toml
plugins: rerunfailures-14.0
collected 6007 items / 6003 deselected / 4 selected                                                                                                                                                                                           

test/test_download.py::TestDownload::test_PlaySuisse PASSED                                                                                                                                                                             [ 25%]
test/test_download.py::TestDownload::test_PlaySuisse_1 PASSED                                                                                                                                                                           [ 50%]
test/test_download.py::TestDownload::test_PlaySuisse_2 PASSED                                                                                                                                                                           [ 75%]
test/test_download.py::TestDownload::test_PlaySuisse_all PASSED                                                                                                                                                                         [100%]

===================================================================================================== 4 passed, 6003 deselected in 9.37s ======================================================================================================
```

Reply:
It's all good, thanks for your work on this